### PR TITLE
Adaptation of @tsing80’s node joining merge tree

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] merge with semver aware protocol
- [x] merge only if overrides option is the same
- [x] merge annotations
- [x] tests

takes:

out2

and makes it:

out10

this particular test has to much noise to verify the performance characteristics. I'll try to clean it up yet.

### CODE REVIEW

**Comment by stefanpenner** [comment]
When layer on top of #34 it is still a 10% -> 15% win.

**Comment by joliss** [comment]
The code looks correct.

For clarification, it's 10-15% gain for merge-trees's own run time, not total build time, yes?

I'm uncertain whether the performance gain is really worth the complexity:

* I generally prefer not to increase complexity for relatively minor(?) performance gains.
* Having an in-memory FS abstraction between nodes could eventually drop the run time of merge-trees to near-zero anyway. Of course we're still quite far away from getting there.
* Even when we merge annotations, I'm guessing it will make understanding the graph a bit harder, because we lose some structure. (I think?)

Do you think it's worth pursuing this further? I'm tending towards "no", but let me know what you think.

**Comment by stefanpenner** [comment]
@joliss actually it was total time for me. As it happens to combine some costly inputTree reading and change detection.

Although I didn't track it it should also improve initial boot.

I can get some more targeted numbers if needed.

Ultimately the win is because ember-cli addons expect some pretty unfortunate and costly merges to occur. 

I believe even with my more holistic cross plugin optimizations I'm currently spiking on this optimization will have some value. Obvious that work is still a ways out so I could be wrong.

**Comment by stefanpenner** [comment]
If you have ideas to reduce the complexity or isolate from it. I'm all ears.

**Comment by joliss** [comment]
Okay, sounds reasonable enough. My only concern is that it might degrade the debugging experience because it changes/obscures the graph structure. But you do all the debugging, so I'll let you be the judge.

Seems like merging annotations plus a simple test case, this should be good to merge.

File: index.js [hunk 1]
```diff
@@ -13,15 +15,38 @@ function BroccoliMergeTrees(inputNodes, options) {
...
+      nodes.push(inputNodes[i]);
+    }
+  }
+
+  return nodes;
+}
+
+BroccoliMergeTrees.prototype[ASSIMILATION_SYMBOL] = function(options, other) {
+  return typeof other[ASSIMILATION_SYMBOL] === 'function' &&
+         options.overwite === other.options.overwite;
```
Comment (joliss) [inline index.js]: We can cast to bool here: `!!options.overwite === !!other.options.overwite`

Comment (stefanpenner) [inline index.js]: good idea

**Comment by stefanpenner** [comment]
> But you do all the debugging, so I'll let you be the judge.

If I run into a case where this seriously sucks, I will fix the debugging.

**Pull request closed without merging**